packages/tpetra/core/src/Tpetra_MultiVector_def.hpp

```diff
@@ -3996,18 +3997,39 @@ namespace Tpetra {
       const LO C_lclNumRows = C_tmp->getLocalLength ();
       const LO C_numVecs = C_tmp->getNumVectors ();
       auto C_lcl = C_tmp->getLocalViewDevice ();
-      auto C_sub = Kokkos::subview (C_lcl,
-                                    std::make_pair (LO (0), C_lclNumRows),
-                                    std::make_pair (LO (0), C_numVecs));
-      const char ctransA = (transA == Teuchos::NO_TRANS ? 'N' :
-                            (transA == Teuchos::TRANS ? 'T' : 'C'));
-      const char ctransB = (transB == Teuchos::NO_TRANS ? 'N' :
-                            (transB == Teuchos::TRANS ? 'T' : 'C'));
+
+      const char ctransA = (transA == NO_TRANS ? 'N' :
+                            (transA == TRANS ? 'T' : 'C'));
+      const char ctransB = (transB == NO_TRANS ? 'N' :
+                            (transB == TRANS ? 'T' : 'C'));
       const impl_scalar_type alpha_IST (alpha);
 
-      ProfilingRegion regionGemm ("Tpetra::MV::multiply-call-gemm");
-      KokkosBlas::gemm (&ctransA, &ctransB, alpha_IST, A_sub, B_sub,
-                        beta_local, C_sub);
+      const size_t zero = 0;
+
+      const bool cond = (transA != NO_TRANS && transB == NO_TRANS);
+      // The first two cases are specializations for inner products between a vector and a multivector
+      if (cond && A_sub.extent(1) == 1) {
+        auto A_sub_sub = subview (A_sub, ALL (), zero);
+        auto C_sub = subview (C_lcl, zero, ALL ());
+        ::Tpetra::Details::ProfilingRegion regionGemm ("Tpetra::MV::multiply-call-gemv");
+        KokkosBlas::gemv(&ctransA, alpha_IST, B_sub, A_sub_sub, beta_local, C_sub);
+
+        if (transA == CONJ_TRANS)
```

Comment:
Doesn't `gemv`'s `ctransA` argument take care of this?  Why do you need another conjugation?

Reply:
We want to compute A^H * B, where A is a vector. We compute B^H * A and then complex conjugate the result.

Reply:
@cgcgcg Why not just compute `B^T * A`?  That way you don't have to double-conjugate.

Reply:
@mhoemmen That does not seem to be correct, no? Cause that way, A does not get conjugated.

Reply:
Could we only do the conjugation for complex scalar types?  I'd rather not launch another kernel in the common case.

Reply:
Shouldn't that already be taken care of by `transA == CONJ_TRANS`? Or should we check whether `CONJ_TRANS` has been passed, even though everything is real-valued?

Reply:
It's legal to pass `CONJ_TRANS` for real-valued types.  In that case, it means the same thing as `TRANS`.

Reply:
> It's legal to pass CONJ_TRANS for real-valued types. In that case, it means the same thing as TRANS.

That is correct.  If you think about it, a real number is just a complex number with the imaginary part being 0.

Reply:
BTW, Thyra has a function for this use case [`real_trans(transp)`](https://trilinos.org/docs/dev/packages/thyra/doc/html/group__Thyra__Op__Vec__fundamental__interfaces__code__grp.html#gae7d8a1c53cf64f6008b62eceb3b93299).  Makes it easy to write operators that are only for real numbers.

Reply:
@bartlettroscoe wrote:
> That is correct. If you think about it, a real number is just a complex number with the imaginary part being 0.

Fun fact: `std::conj(T)` works for real `T`, but always returns a complex number.